### ISSUE TITLE
C interoperability: zero-init and partial initialization for imported C structs?

### ISSUE THREAD

**Issue by floooh** [ziglang/zig#1031]
Hi, I'm attempting to use my 3D-API wrapper [sokol_gfx.h](https://github.com/floooh/sokol) from zig, I got pretty far (importing the C header works, linking with the C implementation source also works), but now I'm hitting a snag when trying to setup 'option-bag structs' that are handed to the sokol_gfx API.

It looks like zig requires that all struct members must be initialized when the struct is 'instantiated', sokol_gfx uses an 'option bag philosophy': non-initialized struct members are assumed to be zero-initialized, which for sokol-gfx indicates that default-values are to be used for such fields. This works nicely with C99-style designated initialization, which sets missing fields to zero.

For instance there are structs like sg_pipeline_desc, which all in all have dozens of members (in embedded structs): https://github.com/floooh/sokol/blob/a3057b13ba38d2d554a7f04e063f6550b2cc8bcb/sokol_gfx.h#L1422

Usually, only a small number of sg_pipeline_desc members are initialized, for instance:

```c
    sg_pipeline pip = {
        .layout = {
            .buffers[0].stride = 28,
            .attrs = {
                [0] = { .name="position", .format=SG_VERTEXFORMAT_FLOAT3 },
                [1] = { .name="color0", .format=SG_VERTEXFORMAT_FLOAT4 }
            }
        },
        .shader = shd,
        .index_type = SG_INDEXTYPE_UINT16,
        .depth_stencil = {
            .depth_compare_func = SG_COMPAREFUNC_LESS_EQUAL,
            .depth_write_enabled = true,
        },
        .rasterizer.cull_mode = SG_CULLMODE_BACK,
    });
```
Currently zig complains about all missing struct members in the imported C structs, and in my case, providing an initialization for *all* struct members manually to 0 wouldn't make sense, the sokol_gfx API would be unusable.

Is there a simple solution to this problem which I missed? Maybe an option for the "@cInclude()" process? 

Also, what's zig's 'best practice' for such 'option-bag' APIs when written in pure zig? As far as I have seen, struct members cannot be assigned a default value, and all members must be initialized, but this gets a bit ugly if there are more than a handful members.

**Comment by andrewrk** [comment]
Here's the current suggestion for this:

```zig
std.mem.set(u8, ([]u8)(&sg_pipeline)[0..1], 0); // set everything to zero
sg_pipeline.shader = shd;
sg_pipeline.index_type = SG_INDEXTYPE_UINT16;
// etc
```

It's a little clumsy because we want to encourage setting every field. May be it's not so bad to explicitly set every field to zero?

**Comment by floooh** [comment]
I'll play around with it and see how it feels. Maybe it makes sense to write a small zig wrapper module for sokol_gfx which hides things like the std.mem.set... Thanks for the quick response :)

**Comment by mrec** [comment]
Did you (Andre) have any luck wrapping this with something that preserves the nice declarative flow of the original? I'm wondering whether this might be another motivating use case for #485

**Comment by floooh** [comment]
@mrec I started to use init-blocks as a workaround but tbh I'm not very happy with it, it feels like a poor replacement for the C99 designated init feature set for arrays:

https://github.com/floooh/sokol-zig/blob/be8b68f4697b072239e18f6af1aba6a6362071b2/examples/cube.zig#L80-L94

E.g. in this case it only looks somewhat acceptable because the array is nested in the `VertexLayoutState` struct which is the result of the init-block but there's not always such a convenient outer struct.

PS: the other issues mentioned above have been solved with default-values in the struct declaration though. The only missing piece is the initialization of arrays.
